xbmc/cores/AudioEngine/Sinks/AESinkALSA.cpp

```diff
@@ -752,7 +756,10 @@ bool CAESinkALSA::InitializeHW(const ALSAConfig &inconfig, ALSAConfig &outconfig
         else if (fmt == SND_PCM_FORMAT_S24 && bits == 24)
           i = AE_FMT_S24NE4;
         else
+        {
+          snd_pcm_hw_params_copy(hw_params, hw_params_copy); // restore hw_params
```

Comment:
this code does not reflect the commit message. you changed behaviour for the unlikely case where bits != fmtBits and format is neither SND_PCM_FORMAT_S32 nor SND_PCM_FORMAT_S24.
please elaborate

Reply:
this code changes behavior to fix an issue where the `snd_pcm_hw_params_set_format` call with S24 or S32 is successful but then bails out to test next format when `bits != 24`, since the `snd_pcm_hw_params_set_format` call was successful the next `snd_pcm_hw_params_set_format` call with S16 fails unless we restore `hw_params` to a state before the successful `snd_pcm_hw_params_set_format`

checking mainline kernel there only seems to be 2 different sbits used where it does not match up with format physical bits, 20-bits and 24-bits (grep `sig_bits` and `msbits` on `sound/`)
I can only find 3 drivers (`pci/intel8x0.c`, `pci/cs4281.c` and `soc/codecs/adau7002.c`) that use 20 significant bits so I imagine not a lot of users being affected by this issue
wanted to include a fix after discovering it when I simulated this state by setting `sig_bits = 20` in kernel

Reply:
seems unconsistent to me. there are more "continues" in this loop but you added the restore only to a singe case.

Reply:
this was the only `continue` after a successful `snd_pcm_hw_params_set_format` call I found, I will move the `snd_pcm_hw_params_copy` call to always run before the `snd_pcm_hw_params_set_format` call, should simplify the code flow

Reply:
that would be great. currently the successful call is hidden as 2nd extression in line 743. I have missed this on first glance.